Add support for immutable collections

rlnt opened the pull request:
# Overview

In this pull request, I added support for the [Kotlinx Collections Immutable](https://github.com/Kotlin/kotlinx.collections.immutable) library.

The goal is to support mapping any iterable to an immutable/persistent set or an immutable/persistent list.

fixes #323 

# Implementation

I decided to create a separate module inside the `modules` directory, similar to how this was done for `kotlinx-datetime`. The reason for this is that immutable collections are an external dependency, not being part of the Kotlin stdlibs.

Because the library supports all platforms the `mappie-api` supports as well, I decided to support all platforms for this module too.

Right now, this pull request features a mapper to convert an `Iterable` to an `ImmutableList` and another mapper to convert an `Iterable` to an `ImmutableSet`. The same applies to mappers of the persistent collections.

# Caveats

This pull request is currently marked as draft because it's not fully ready. ~~the implementation doesn't fully work yet. I had to fix a few things that need discussing because these are unrelated to the mappers themselves.~~

I had to move the `PredefinedMappieProvider` interface to the `commonMain` source set. This might be a Gradle issue, but since I could implement all mappers in the `commonMain` source set, I wanted to have the interface implementation there as well. The problem was that the interface couldn't be found on the classpath. After moving the interface implementation to the `jvmMain` source set, the interface was found. However, this led to a build failure when trying to publish to Maven local, because the `ExternalDefinitionsCollector` inside the `compiler-plugin` module couldn't find the implementation anymore.
The `kotlinx-datetime` integration has the `PredefinedMappieProvider` implementation inside its `commonMain` source set, and I have no idea why this works. The `mappie-api` module has the implementation in its `jvmMain` source set, but the `ExternalDefinitionsCollector` compiles just fine.
When I do any of that, the build fails. The only solution I found was to move the interface to the common source set. In my opinion, this would be more accurate as well if other modules want to implement this in common too.

~~Second, I had to adjust the `MappieDefinitions` erasure logic. I am not too familiar with Kotlin internals and how the immutable collections work, but when testing the changes in my personal project that tries to map a `List` to an `ImmutableList`, I got an error about a wildcard generic being in use, which was previously unhandled in the internal mappie logic. The new logic treats `StarProjection`s as `Any`. This could be improved to retrieve the upper generic bound, but I am not experienced enough to improve this further.~~

~~Third, this implementation still does not work. I am getting a compiler error when trying to map a `List` to an `ImmutableList`.~~

# Todo

~~There is no implementation for `ImmutableMap`. I don't know how mappie currently treats maps, but this may need some special implementation.~~ Adding immutable map support is out of scope for this pull request.

~~Since `ImmutableList` and `ImmutableSet` are strict types, we should consider adding built-in support for `PersistentList` and `PersistentSet` as well. These types inherit the immutable counterpart but offer utility functions to modify the collection by returning a copy with the modifications.~~

The website needs to be updated to list the new immutable collection integration.

The new module requires publishing. Adjustments to the workflow are necessary.

~~Proper tests need to be implemented, but I need to make this compile first.~~

# Side Notes

~~The documentation for `kotlinx-datetime` is really hidden and wrong. The only appearance of it is on the "Built-in Mappers" page at the bottom. It's hard to find. Additionally, it mentions that you have to include the dependency via `tech.mappie.api:module-kotlinx-datetime`, but the dependency is actually called `tech.mappie:module-kotlinx-datetime`.~~

rlnt commented:
Hmm, no. The star projection issue seems to originate from something else. I can reproduce this without my PR when changing from 1.6.1 to 2.0.0. I opened another issue submission for this.
https://github.com/Mr-Mappie/mappie/issues/331

stefankoppier commented:
> If I manage to fix the underlying issue, I will try to get rid of this change to see if it works without it. Handling star projections could be a good improvement for the future anyway.

I will give the bug report a look, and will make some tests and fixes.

> As far as I understand, Mappie loads modules automatically if you are loading the dependency, right? Maybe it isn't correctly loaded in my case.

This is the case for kotlinx-datetime, yes. It seems to be added in `MappieCompilerPluginRegistrar#37` in your branch, maybe the Regex doesn't match. If we add a Gradle test similar to `KotlinxDateTimeTest` we should be able to debug if the regex is matched or not (I will do this right now to help you forward).

> Is anything else required to load a separate module?

If I remember correctly, no. But I will come back to you :)

rlnt commented:
I excluded the star projection change from this PR and rebased the branch on the latest state of `main`. I also added mappers for the persistent collection types, as well as plenty of test coverage.

The description of the pull request was updated to reflect the latest changes. The last remaining problem is the location of the `BuiltInMappieProvider` interface. It's still part of `commonMain` because I was not able to move it back to `jvmMain` without compilation failing. I still have no clue why the `kotlinx-datetime` module even works.

stefankoppier commented:
LGTM, thanks!
